### DISCUSSION TITLE
configEntries not containing overridden values

### DISCUSSION THREAD

**Discussion by dajinchu** [vikejs/vike#1027] [Discussion] [answered]
I would like to be able to get overridden config values in configEntires. 

Use case: I have `documentProps` that contains `viewport`, which I'd like to set a default in `renderer/config.h.ts` while being able to override documentProps at a per-page level.

I understand this behavior works by default if I pulled `viewport` out into its own config and set it in `passToClient`, but I'd rather have configs logically grouped if possible, then when accessing `documentProps` I can shallow merge the overridden values.

Is configEntries not meant to include overridden values?

**Comment by brillout** [comment]
Yes it's on the radar.

> Is configEntries not meant to include overridden values?

It only includes values that are loaded. Overriden values aren't loaded (to save loading time).

That's why a new `meta` property is needed:

```js
export default {
  meta: {
    viewport: {
      cumulative: true
      /* Alternative naming:
      aggregate: true
      override: false
      overridable: false
      */
    }
  }
}
```

> but I'd rather have configs logically grouped if possible, then when accessing `documentProps` I can shallow merge the overridden values.

I'm not sure I understand your use case, do you mind elaborating?

**Reply by dajinchu** [reply]
Makes a lot of sense. cumulative would solve this for me. I just want to be able to do:

```typescript
# renderer/config.h.ts

export default {
  documentProps: {
    viewport: "whatever"
  }
}

# pages/mypage/config.h.ts

export default {
  documentProps: {
    title: "my title"
  }
}
```

and have the documentProps be merged, so `mypage` gets the default viewport and the custom title. But at the moment, declaring a title overwrites the viewport. Technically I could solve this without `cumulative` by moving viewport out of documentProps:

```typescript
# renderer/config.h.ts

export default {
  viewport: "whatever"
}

# pages/mypage/config.h.ts

export default {
  documentProps: {
    title: "my title"
  }
}
```

But it feels illogical. I'd rather keep all document metadata in documentProps.

**Reply by brillout** [reply]
I'd actually recommend to sperate both as it's slightly cleaner since both configs have different logic. (FYI the `vike-*` packages will have several flat metadata configs instead of a single `documentProps`.)

FYI I'm also considering nested configs for better organization:

```js
export default {
  meta: {
    ['documentProps.viewport']: {
      env: 'server-only'
    },
    ['documentProps.title']: {
      env: 'server-and-client'
    }
  }
}
```

Giving you the best of both words.

Note that both `cumulative` and nested configs aren't the highest priority at the moment. But let me know if it becomes a blocker or too much of an annoyance and I'll implement it. (The `cumulative` option should be fairly quick to implement.) Especially given that you're a sponsor, so definitely keep me in the loop.

**Reply by dajinchu** [reply]
Great to know, thanks. I will consider moving configs out

**Reply by dajinchu** [reply]
@brillout Would it be possible to get `cumulative` bumped somewhat in priority? I am seeing more and more need for it. For example, our app adds a `scripts` array config to add third party scripts. Certain scripts are required for every page in the app, some only for certain subpaths, and some only for very specific pages. It's getting a little messy importing from `/renderer` and merging arrays in userland. Thanks!

**Reply by brillout** [reply]
Yes, we can bump it. Although note that there is an internal refactor in progress tha needs to be finished before `cumulative` can be implemented. So it may take a little while. I'll tentativelly implement it end of this week. Is that an acceptable ETA?

**Reply by brillout** [reply]
To clarify that we're on the same page: you want all your `scripts` arrays to be cumulative, correct? In other words: you don't want/need some `scripts` array to override some other `scripts` array, right?

**Reply by dajinchu** [reply]
> To clarify that we're on the same page: you want all your `scripts` arrays to be cumulative, correct? In other words: you don't want/need some `scripts` array to override some other `scripts` array, right?

Yeah they'd all be cumulative. Thanks a ton. A week or so is great.

**Reply by brillout** [reply]
> A week or so is great.

👍 I'll keep you updated how that goes. The duration of this internal refactor isn't easy to predict.

**Reply by brillout** [reply]
The new internal data structure is implemented (unlocks cumulative and many other things such as reduced cilent-side KBs, Single Route File, ...).

I expect it to now only be a matter of days. ETA sometime next week.

**Reply by dajinchu** [reply]
Awesome! super excited

**Reply by brillout** [reply, marked as answer]
Done and pre-released in `0.4.140-commit-710acc9`. I hope the delay was ok for you (as always: let me know if something is a blocker).

If you can wait a little bit more, I'll be polishing a couple of internals and then release a new minor. You can use the pre-release in the meantime.

```js
// +config.h.js
export default {
  meta: {
    viewport: {
      cumulative: true
    }
  },
  viewport: [500]
}
```

Cumulative configs have a couple of minor limitations, let me know if one of these are problematic for you. (I expect the limitations to not be a problem for the vast majority of cumulative configs.)

**Reply by brillout** [reply]
Released in `0.4.141`.

**Reply by brillout** [reply]
@dajinchu `0.4.169` contains a breaking change about custom configs that are cumulative.

```js
// /pages/+config.js

export default {
  someCumulativeConfig: ['a']
}
```
```js
// /pages/some-page/+config.js

export default {
  someCumulativeConfig: ['b']
}
```

Breaking change:

```js
// <= 0.4.168
console.log(pageContext.config.someCumulativeConfig) // prints ['a', 'b']

// >= 0.4.169
console.log(pageContext.config.someCumulativeConfig) // prints [['a'], ['b']]
```

Thus make sure to apply this change:

```diff
- const val = pageContext.config.someCumulativeConfig
+ // @ts-ignore
+ const val: string[] = pageContext.config.someCumulativeConfig.flat()
```

The `@ts-ignore` is because the TS type of `pageContext.config` is now broken for cumulative configs, which is something that is going to be fixed in the next release.

This breaking change is important for many reasons (and given that only few use it didn't seem to justify bumping to `0.5`).

As always let me know if you run into any issues, and thank you a lot for sponsoring  :green_heart: it makes a massive difference.

**Reply by brillout** [reply]
> The `@ts-ignore` is because the TS type of `pageContext.config` is now broken for cumulative configs, which is something that is going to be fixed in the next release.

@dajinchu Done: see `Vike.ConfigResolved` at https://vike.dev/meta#typescript. Make sure to upgrade to `0.4.172`.
